rfcs/2020-05-25-more-usable-logevents.md

```diff
@@ -0,0 +1,223 @@
+		[true, null, true],
+		{
+			"basic": true,
+			"buddy": 1.0, // Maps with multiple values are a bit more complex.
+		}.
+	],
+	"map": {
+		"basic": true,
+		"list": [true, null, true],
+		"map": {
+			"basic": true,
+			"buddy": -1,
+		}
+	},
+}
+
+```
+
+## Dogma and Intent
+
+Vector has embraced the idea of **Paths**, which are analogous to CSS Selectors or `jq` selectors, that can represent an arbitrary selection of value(s). It is irrelevant to the user how the `LogEvent` is internally structured, so long as they can utilize these paths to work with a `LogEvent`.
+
+The `LogEvent` variant used under a two primary intents:
+
+- **Programmatically, through our internal code.** When used in this way, the `LogEvent` is being interacted with through static, safe interfaces and we can utilize the full type system available to us.
+- **Through configuration or FFI via knobs or scripts.** When used this way, the configuration format (at this time, TOML) may impact how aesthetic or expressive our configuration is. For example, TOML values like `a.b.c` as defined in a configuration. This is TOML table syntax, and we can't just magically treat it like a `Path`, a conversion step is necessary.
+
+## Understanding `LogEvent` in Vector 0.9.0
+
+Before we explore how to improve `LogEvent`, we must make sure we're familiar with the current one! Let's review the structure and some examples of lifetimes of this structure.
+
+### `LogEvent` Internal Structure
+
+It's current implementation meets our needs currently from several ergonomics angles (notably the flattening code), but it we must twist it in several painful ways to work with it (notably buffers to flatmaps to `BTree` style APIs).
+
+```rust
+//! src/event/mod.rs
+#[derive(PartialEq, Debug, Clone)]
+pub struct LogEvent {
+    fields: BTreeMap<String, Value>,
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub enum Value {
+    Null,
+    Boolean(bool),
+    Integer(i64),
+    Float(f64),
+    Timestamp(DateTime<Utc>),
+    Bytes(Bytes),
+    Array(Vec<Value>),
+    Map(BTreeMap<String, Value>)
+}
+```
+
+The current `LogEvent` type's internal data model bears semblance to the [`serde_json::Value::Value::Object(Map<String, Value>)`](https://docs.serde.rs/serde_json/value/enum.Value.html) variant.
+
+```rust
+//! serde_json::value::Value
+pub enum Value {
+    Null,
+    Bool(bool),
+    Number(Number),
+    String(String),
+    Array(Vec<Value>),
+    Object(Map<String, Value>),
+}
+```
+
+Differences are:
+
+- `Event` uses a `BTreeMap` , where `serde_json`'s `Map` type is only `BTreeMap` when the `preserve_order` feature flag is turned off. We do not enable this, feature, so they are equivalent.
+- A `Bytes` type in our `Value` type exists because `serde_json::value::Value` and most JSON implementations support this type through the `String` type, which isn't ideal for applying `Bytes` related functions on. We hold this special `Bytes` variant to support better in-pipeline processing for Vector.
```

Comment:
"applying … on" ⇒ "applying … to" or even just drop the "on". Might want to mention the lossy nature of Bytes-to-String conversions here or below.

Reply:
We've discussing removing those conversions now